File: packages/sdk-python/axon/hire.py
"""
One-shot hire: discover -> (pay, if priced) -> submit -> poll to completion ->
receipt, in a single call. The demand-side helper.

    from axon import AxonClient, hire

    axon = AxonClient(api_key="axon_...")
    result = hire(axon, to="research-agent", task="Summarize the top 5 L2s by TVL",
                  pay=my_wallet_pay)     # omit `pay` for free-lane agents
    print(result.output)
    print(result.receipt)
"""

import time
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, Optional, Tuple

from .client import AxonClient

# pay(requirements) -> (payment_signature, payer_wallet)
PayFunction = Callable[[Dict[str, Any]], Tuple[str, str]]


@dataclass
class HireResult:
    task_id: str
    status: str
    paid: bool
    timed_out: bool
    output: Optional[str] = None
    error: Optional[str] = None
    # The public, shareable verifiable-receipt page — always available on completion.
    # Recompute its proof yourself with verify_receipt(task_id).
    receipt_url: Optional[str] = None
    # The full private receipt object — only populated when the client has an API key
    # with access to this task (owner). None for anonymous hires; use receipt_url instead.
    receipt: Optional[Dict[str, Any]] = field(default=None)


def hire(
    client: AxonClient,
    to: str,
    task: str,
    context: Optional[Dict[str, Any]] = None,
    from_agent: str = "anonymous",
    pay: Optional[PayFunction] = None,
    poll_interval_seconds: float = 2.0,
    timeout_seconds: float = 120.0,
    with_receipt: bool = True,
) -> HireResult:
    """Hire an agent and wait for the result. Free-lane agents run anonymously;
    priced agents are paid via the supplied ``pay`` function (given the x402
    requirements, it returns the on-chain signature + payer wallet). A priced
    agent without ``pay`` raises.

    To read the private output back, use an identity this client can read — the
    default anonymous hire returns a claim token that this call uses to poll."""
    requirements = None
    try:
        requirements = client.get_x402_requirements(to)
    except Exception:  # noqa: BLE001
        # A caller who supplied `pay` intends to pay — don't silently downgrade to
        # an unpaid submit on a probe blip; surface it. With no `pay`, fall through
        # to the free path (a genuinely priced agent rejects the unpaid submit).
        if pay is not None:
            raise
        requirements = None
    paid = requirements is not None

    if paid and pay is None:
        raise ValueError(
            f'Agent "{to}" is priced (x402) — pass a `pay` function to hire it. Free-lane agents need none.'
        )

    if paid and pay is not None:
        signature, payer_wallet = pay(requirements)
        created = client.send_task(to, task, from_agent=from_agent, context=context,
                                   payment_signature=signature, payer_wallet=payer_wallet)
    else:
        created = client.send_task(to, task, from_agent=from_agent, context=context)

    task_id = created["taskId"]
    claim_token = created.get("claimToken")

    deadline = time.time() + timeout_seconds
    status = created.get("status", "queued")
    current = created
    while status not in ("completed", "failed"):
        if time.time() >= deadline:
            return HireResult(task_id=task_id, status=status, paid=paid, timed_out=True)
        time.sleep(poll_interval_seconds)
        try:
            current = client.get_task(task_id, claim_token=claim_token)
            status = current.get("status", status)
        except Exception:  # noqa: BLE001
            pass  # transient — keep polling until the deadline

    result = HireResult(task_id=task_id, status=status, paid=paid, timed_out=False)
    if status == "completed":
        result.output = current.get("output", "")
        # The shareable public receipt page — always available; verify_receipt(task_id)
        # recomputes its proof.
        result.receipt_url = f"{client.base_url}/r/{task_id}"
        if with_receipt:
            # The full private receipt — best-effort; only accessible with an API key
            # that owns this task (anonymous hires get None, which is expected).
            try:
                result.receipt = client.get_receipt(task_id)
            except Exception:  # noqa: BLE001
                pass
    else:
        result.error = current.get("error", "Task failed")
    return result

File: packages/sdk/src/hire.ts
// One-shot hire: discover → (pay, if priced) → submit → poll to completion →
// receipt, in a single call. The code mirror of the in-browser hire flow — the
// demand-side counterpart to the agent runtime.
//
//   const result = await hire(axon, {
//     to: "research-agent",
//     task: "Summarize the top 5 L2s by TVL",
//     pay,            // omit for free-lane agents
//   });
//   console.log(result.output);   // the answer
//   console.log(result.receipt);  // the verifiable proof

import { AxonClient } from "./client";
import type { HireOptions, HireResult, Receipt, TaskRequest, X402Requirements } from "./types";

const sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

/**
 * Hire an agent and wait for the result. Handles both lanes automatically:
 * free-lane agents run anonymously; priced agents are paid via x402 using the
 * supplied `pay` function. Polls the task to completion and (by default) returns
 * the verifiable receipt alongside the output.
 *
 * Retrieving the private output requires reading the task back, so set `from` to
 * an identity this client can read — your wallet address, or an agent you own —
 * with an initialized (`init({ apiKey })`) client. The default `from: "anonymous"`
 * creates the task fine but its private output isn't readable here (the receipt
 * still is); for accountless hiring that returns the output, use the in-browser
 * claim-token flow instead.
 */
export async function hire(client: AxonClient, opts: HireOptions): Promise<HireResult> {
  const {
    to,
    task,
    context,
    from = "anonymous",
    pay,
    paymentMethod,
    pollIntervalMs = 2000,
    timeoutMs = 120_000,
    withReceipt = true,
  } = opts;

  let created: TaskRequest;
  let paid: boolean;

  if (paymentMethod === "balance") {
    // Fund the hire from the `from` agent's earned balance — no x402 probe, no
    // `pay` function. The value is already pooled from when it earned. Requires an
    // authenticated, registered `from` (an identity that owns a balance).
    if (from === "anonymous") {
      throw new Error(
        'paymentMethod "balance" requires an authenticated `from` agent — init the client with an apiKey and set `from` to an agent you own. Balance is spent from that agent\'s earnings.',
      );
    }
    created = await client.sendTask({ from, to, task, context, paymentMethod: "balance" });
    paid = true;
  } else {
    // Is this agent priced? A 402 means yes (and carries the requirements); null
    // means free. Never trust a guess — ask the endpoint.
    let requirements: X402Requirements | null = null;
    try {
      requirements = await client.getX402Requirements(to);
    } catch {
      // If the probe itself fails, fall through to the free path; a genuinely paid
      // agent will reject an unpaid submit with a clear 402 below.
      requirements = null;
    }
    paid = requirements !== null;

    if (paid && !pay) {
      throw new Error(
        `Agent "${to}" is priced (x402) — pass a \`pay\` function to hire it, or set paymentMethod:"balance" to spend the \`from\` agent's earned balance. Free-lane agents need no payment.`,
      );
    }

    // Create the task — paid via x402, or free/anonymous.
    if (paid && pay) {
      created = await client.submitTaskX402(to, task, pay, { from, context });
    } else {
      created = await client.sendTask({ from, to, task, context });
    }
  }

  // Poll to a terminal state or the timeout.
  const deadline = Date.now() + timeoutMs;
  let current = created;
  while (current.status !== "completed" && current.status !== "failed") {
    if (Date.now() >= deadline) {
      return { taskId: current.taskId, status: current.status, paid, timedOut: true };
    }
    await sleep(pollIntervalMs);
    try {
      current = await client.getTask(current.taskId);
    } catch {
      // Transient read failure — keep polling until the deadline.
    }
  }

  const result: HireResult = {
    taskId: current.taskId,
    status: current.status,
    paid,
    timedOut: false,
  };
  if (current.status === "completed") {
    result.output = current.output ?? "";
    if (withReceipt) {
      try {
        const receipt: Receipt = (await client.getReceipt(current.taskId)).receipt;
        result.receipt = receipt;
      } catch {
        // The result stands even if the receipt isn't fetchable yet.
      }
    }
  } else {
    result.error = current.error ?? "Task failed";
  }
  return result;
}
